test/functional/p2p-fingerprint.py

```diff
@@ -0,0 +1,158 @@
+
+    # Check whether last block header received from node has a given hash
+    def last_header_equals(self, expected_hash, node):
+        headers_msg = node.last_message.get("headers")
+        return (headers_msg and
+                headers_msg.headers and
+                headers_msg.headers[0].rehash() == expected_hash)
+
+    # Checks that orphan blocks timestamped more than a month ago are not served
+    # by the node while recent orphan blocks and old active chain blocks are.
+    # This does not currently test that orphan blocks timestamped within the
+    # last month but that have over a month's worth of work are also withheld.
+    def run_test(self):
+        node0 = NodeConnCB()
+
+        connections = []
+        connections.append(NodeConn('127.0.0.1', p2p_port(0), self.nodes[0], node0))
+        node0.add_connection(connections[0])
+
+        NetworkThread().start()
+        node0.wait_for_verack()
+
+        # Set node time to 60 days ago
+        self.nodes[0].setmocktime(int(time.time()) - 60 * 24 * 60 * 60)
+
+        # Generating a chain of 10 blocks
+        block_hashes = self.nodes[0].generate(nblocks=10)
+
+        # Create longer chain starting 2 blocks before current tip
+        height = len(block_hashes) - 2
+        block_hash = block_hashes[height - 1]
+        block_time = self.nodes[0].getblockheader(block_hash)["mediantime"] + 1
+        new_blocks = self.build_chain(5, block_hash, height, block_time)
+
+        # Force reorg to a longer chain
+        node0.send_message(msg_headers(new_blocks))
+        node0.wait_for_getdata()
+        for block in new_blocks:
+            node0.send_and_ping(msg_block(block))
+
+        # Check that reorg succeeded
+        assert_equal(self.nodes[0].getblockcount(), 13)
+
+        orphan_hash = int(block_hashes[-1], 16)
+
+        # Check that getdata request for orphan block succeeds
+        self.send_block_request(orphan_hash, node0)
+        test_function = lambda: self.last_block_equals(orphan_hash, node0)
+        wait_until(test_function, timeout=3)
+
+        # Check that getheader request for orphan block header succeeds
+        self.send_header_request(orphan_hash, node0)
+        test_function = lambda: self.last_header_equals(orphan_hash, node0)
+        wait_until(test_function, timeout=3)
+
+        # Longest chain is extended so orphan is much older than chain tip
+        self.nodes[0].setmocktime(0)
+        tip = self.nodes[0].generate(nblocks=1)[0]
+        assert_equal(self.nodes[0].getblockcount(), 14)
+
+        # Send getdata & getheaders to refresh last received getheader message
+        block_hash = int(tip, 16)
+        self.send_block_request(block_hash, node0)
+        self.send_header_request(block_hash, node0)
+        node0.sync_with_ping()
+
+        # Request for very old orphan block should now fail
+        self.send_block_request(orphan_hash, node0)
+        time.sleep(3)
```

Comment:
Also use `wait_until` instead of fixed sleep?

Reply:
That doesn't work. We're waiting to make sure a change does not happen.